RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc

```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+                               momentum * t.barycenter().unit().y(),
+                               momentum * t.barycenter().unit().z(),
+                               t.raw_energy());
+    tmpCandidate.setP4(p4);
+    resultCandidates->push_back(tmpCandidate);
+  }
+
+  // Charged Particles
+  for (unsigned i = 0; i < trackstersTRKEM.size(); ++i) {
+    auto mergedIdx = indexInMergedCollTRKEM[i];
+    if (!usedTrackstersMerged[mergedIdx]) {
+      const auto &t = trackstersTRKEM[i];  //trackster
+      auto trackIdx = t.seedIndex();
+      auto const &track = tracks[trackIdx];
+      if (!usedSeeds[trackIdx] and t.raw_energy() > 0) {
+        usedSeeds[trackIdx] = true;
+        usedTrackstersMerged[mergedIdx] = true;
+
+        std::vector<int> trackstersTRKwithSameSeed;
+        std::vector<int> trackstersTRKEMwithSameSeed;
+
+        for (const auto &tracksterIterationPair : seedToTracksterAssociator[trackIdx]) {
+          if (tracksterIterationPair.first != mergedIdx and !usedTrackstersMerged[tracksterIterationPair.first] and
+              trackstersMergedHandle->at(tracksterIterationPair.first).raw_energy() > 0.) {
+            if (tracksterIterationPair.second == TracksterIterIndex::TRKEM) {
+              trackstersTRKEMwithSameSeed.push_back(tracksterIterationPair.first);
+            } else if (tracksterIterationPair.second == TracksterIterIndex::TRK) {
+              trackstersTRKwithSameSeed.push_back(tracksterIterationPair.first);
+            }
+          }
+        }
+
+        float tracksterTotalRawPt = t.raw_pt();
+        std::vector<int> haloTrackstersTRKIdx;
+        bool foundCompatibleTRK = false;
+
+        for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+          usedTrackstersMerged[otherTracksterIdx] = true;
+          tracksterTotalRawPt += trackstersMergedHandle->at(otherTracksterIdx).raw_pt();
+
+          // Check the X,Y,Z barycenter and merge if they are very close (halo)
+          if ((t.barycenter() - trackstersMergedHandle->at(otherTracksterIdx).barycenter()).mag2() <
+              halo_max_distance2_) {
+            haloTrackstersTRKIdx.push_back(otherTracksterIdx);
+
+          } else {
+            foundCompatibleTRK = true;
+          }
+        }
+
+        //check if there is 1-to-1 relationship
+        if (trackstersTRKEMwithSameSeed.empty()) {
+          if (foundCompatibleTRK) {
+            TICLCandidate tmpCandidate;
+            tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, mergedIdx));
+            tmpCandidate.setCharge(track.charge());
+            tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, trackIdx));
+            tmpCandidate.setPdgId(211 * track.charge());
+            for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+              tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, otherTracksterIdx));
+            }
+            float p = tracksterTotalRawPt * cosh(t.barycenter().eta());
```

Comment:
You are using consistently in sevaral places `p = pt * cosh(vec.eta())`: just for my own education, is it faster than (e.g.) `p = pt * sqrt(1+vec.z()*vec.z()/(vec.x()*vec.x()+(vec.y()*vec.y())` ?

Reply:
of course not, thanks!